database/src/backend/heed/database.rs

```diff
@@ -26,72 +26,127 @@ use crate::{
-    tx_rw: &'env mut heed::RwTxn<'env>,
+    pub(super) tx_rw: &'tx mut heed::RwTxn<'env>,
+}
+
+//---------------------------------------------------------------------------------------------------- Shared functions
+// FIXME: we cannot just deref `HeedTableRw -> HeedTableRo` and
+// call the functions since the database is held by value, so
+// just use these generic functions that both can call instead.
+
+/// Shared generic `get()` between `HeedTableR{o,w}`.
+#[inline]
+fn get<'tx, T: Table>(
+    db: &'_ HeedDb<T::Key, T::Value>,
+    tx_ro: &'tx heed::RoTxn<'_>,
+    key: &T::Key,
+) -> Result<impl Borrow<T::Value> + 'tx, RuntimeError> {
+    match db.get(tx_ro, key) {
+        Ok(Some(value)) => Ok(value),
+        Ok(None) => Err(RuntimeError::KeyNotFound),
+        Err(e) => Err(e.into()),
+    }
+}
+
+/// Shared generic `get_range()` between `HeedTableR{o,w}`.
+#[inline]
+#[allow(clippy::needless_pass_by_value, clippy::trait_duplication_in_bounds)]
+fn get_range<'tx, T: Table, Range>(
+    db: &'tx HeedDb<T::Key, T::Value>,
+    tx_ro: &'tx heed::RoTxn<'_>,
+    range: Range,
+) -> Result<impl Iterator<Item = Result<impl Borrow<T::Value> + 'tx, RuntimeError>>, RuntimeError>
+where
+    Range: RangeBounds<T::Key> + RangeBounds<&'tx T::Key> + 'tx,
+{
+    /// TODO
+    struct Iter<'tx, T: Table> {
+        /// TODO
+        iter: heed::RoRange<'tx, StorableHeed<T::Key>, StorableHeed<T::Value>>,
+    }
+
+    // TODO
+    impl<'tx, T: Table> Iterator for Iter<'tx, T> {
+        type Item = Result<&'tx T::Value, RuntimeError>;
+        fn next(&mut self) -> Option<Self::Item> {
+            // TODO
+            self.iter
+                .next()
+                .map(|result| result.map(|value| value.1).map_err(RuntimeError::from))
+        }
+    }
+
+    // TODO
+    match db.range(tx_ro, &range) {
+        Ok(iter) => Ok(Iter::<'_, T> { iter }),
+        Err(e) => Err(RuntimeError::from(e)),
+    }
 }
 
 //---------------------------------------------------------------------------------------------------- DatabaseRo Impl
-impl<T: Table> DatabaseRo<T> for HeedTableRo<'_, T> {
-    fn get(&self, key: &T::Key) -> Result<&T::Value, RuntimeError> {
-        todo!()
+impl<'tx, T: Table> DatabaseRo<'tx, T> for HeedTableRo<'tx, T> {
+    #[inline]
+    fn get<'a>(&'a self, key: &'a T::Key) -> Result<impl Borrow<T::Value> + 'a, RuntimeError> {
+        get::<T>(&self.db, self.tx_ro, key)
     }
 
-    fn get_range<'a>(
+    #[inline]
+    #[allow(clippy::trait_duplication_in_bounds)]
+    fn get_range<'a, Range>(
         &'a self,
-        key: &'a T::Key,
-        amount: usize,
-    ) -> Result<impl Iterator<Item = &'a T::Value>, RuntimeError>
+        range: Range,
+    ) -> Result<impl Iterator<Item = Result<impl Borrow<T::Value> + 'a, RuntimeError>>, RuntimeError>
     where
-        <T as Table>::Value: 'a,
+        Range: RangeBounds<T::Key> + RangeBounds<&'a T::Key> + 'a,
```

Comment:
just make this RangeBounds<T::Key> IMO, this will only be used on keys with simple types (u64)

Reply:
This is due to https://docs.rs/redb/latest/redb/trait.ReadableTable.html#tymethod.range.

- `heed` wants `Range: RangeBounds<T::Key>`
- `redb` wants `Range: RangeBounds<impl Borrow<T::Key>>`

The current `redb` code is `&T::Key`. If we switch our key to `Cow<'_, T::Key>` too, our range has to `impl Borrow<Cow<'_, T::Key>>`.

I guess the input would be:
```rust
let range = Cow::<u64>::Owned(1)..Cow::<u64>::Owned(10);
```

This does create a incompatibility when switching backends since `heed` wants a `u64` and `redb` wants a `Cow<'_, u64>`. We could switch it so `heed` also uses `Cow` internally I guess...?

Reply:
doesn't T::Key impl Borrow for it self? https://doc.rust-lang.org/std/borrow/trait.Borrow.html#impl-Borrow%3CT%3E-for-T

Reply:
To be more clear:
- `heed` wants `Range: RangeBounds<T::Key>` (`heed` sees `&'a T` as the key)
- `redb` wants `Range: RangeBounds<Cow<'a, T::Key>> + 'a` (`redb` sees `Cow<'_, T>` as the key)

We could switch `heed` so it sees `Cow<'_, T>` instead of `&'a T` as the key type, but means we're wrapping all key/values in it during serialization (not bad, still gets serialized as T). It means our ranges will be like the above though...

Or we can make `redb` accept `RangeBounds<T::Key>` then in `redb`'s specific `get_range()` we somehow transform:
```rust
RangeBounds<T::Key>
// ...into...
RangeBounds<Cow<'?, T::Key>>
```
is this possible?